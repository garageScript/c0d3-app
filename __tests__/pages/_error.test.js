import * as React from 'react'
import { render } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import MyError from '../../pages/_error'
import * as Sentry from '@sentry/browser'
jest.spyOn(Sentry, 'captureException')

describe('_error page', () => {
  test('Should return Error 500 component', async () => {
    const { container } = render(
      <MockedProvider addTypename={false}>
        {MyError.getLayout(
          <MyError
            code={500}
            hasGetInitialPropsRun={true}
            err={new Error('test error')}
          />
        )}
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
    expect(Sentry.captureException).not.toHaveBeenCalled()
  })
  test('Should capture error when component is loading', async () => {
    const { container } = render(
      <MockedProvider addTypename={false}>
        {MyError.getLayout(
          <MyError
            code={505}
            hasGetInitialPropsRun={false}
            err={new Error('test error')}
          />
        )}
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
    expect(Sentry.captureException).toHaveBeenCalled()
  })
  test('Should capture browser error', async () => {
    const input = {
      err: { statusCode: 501 },
      asPath: '/pages/someComponent'
    }
    expect(await MyError.getInitialProps(input)).toEqual({
      code: 501,
      hasGetInitialPropsRun: true
    })

    expect(Sentry.captureException).toHaveBeenCalledWith({ statusCode: 501 })
  })
  test('Should capture server error', async () => {
    const input = {
      res: { statusCode: 505 },
      asPath: '/pages/someComponent'
    }
    expect(await MyError.getInitialProps(input)).toEqual({
      code: 505,
      hasGetInitialPropsRun: true
    })

    expect(Sentry.captureException).toHaveBeenCalledWith(
      new Error(
        `_error.js getInitialProps missing data at path: /pages/someComponent`
      )
    )
  })
  test('Should capture unexpected error', async () => {
    const input = {
      asPath: '/pages/someComponent'
    }
    expect(await MyError.getInitialProps(input)).toEqual({
      code: 404,
      hasGetInitialPropsRun: true
    })

    expect(Sentry.captureException).toHaveBeenCalledWith(
      new Error(
        `_error.js getInitialProps missing data at path: /pages/someComponent`
      )
    )
  })
})
