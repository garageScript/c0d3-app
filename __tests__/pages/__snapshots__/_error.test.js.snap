// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_error page Should capture error when component is loading 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="d-flex col-sm-3 align-items-center justify-content-center"
        >
          <div
            class="text-center mt-3"
          >
            <div
              style="color: rgb(41, 41, 41); font-size: 1.5rem;"
            >
              <strong>
                Internal server error
              </strong>
            </div>
            <div
              style="color: rgb(41, 41, 41); font-size: 1rem;"
            />
            <a
              class="btn btn-primary py-3 px-5 mt-3"
              href="/"
            >
              <h3
                class="font-weight-bold"
              >
                Back
              </h3>
            </a>
          </div>
        </div>
        <div
          class="mb-3 col-sm-9"
        >
          <img
            height="1200"
            layout="responsive"
            src="/assets/errors/500.svg"
            width="1200"
          />
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;

exports[`_error page Should return Error 500 component 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="d-flex col-sm-3 align-items-center justify-content-center"
        >
          <div
            class="text-center mt-3"
          >
            <div
              style="color: rgb(41, 41, 41); font-size: 1.5rem;"
            >
              <strong>
                Internal server error
              </strong>
            </div>
            <div
              style="color: rgb(41, 41, 41); font-size: 1rem;"
            />
            <a
              class="btn btn-primary py-3 px-5 mt-3"
              href="/"
            >
              <h3
                class="font-weight-bold"
              >
                Back
              </h3>
            </a>
          </div>
        </div>
        <div
          class="mb-3 col-sm-9"
        >
          <img
            height="1200"
            layout="responsive"
            src="/assets/errors/500.svg"
            width="1200"
          />
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;
